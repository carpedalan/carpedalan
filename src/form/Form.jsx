import React, { createContext, useEffect, useReducer } from 'react';
import { arrayOf, node, shape, func, oneOfType, string } from 'prop-types';
import get from 'lodash/get';
import unset from 'lodash/unset';
import isUndefined from 'lodash/isUndefined';
import set from 'lodash/set';
import isEmpty from 'lodash/isEmpty';

import log from '../utils/log';

export const FormContext = createContext({});
const { Provider: FormProvider } = FormContext;

const initialFormReducerState = {
  values: {},
  errors: {},
  meta: {},
};

const reducer = (
  state,
  { type, payload, payload: { field, value, error } = {} },
) => {
  log.info(type, payload);
  switch (type) {
    case 'CHANGE':
      set(state.values, field, value);
      return {
        ...state,
        meta: {
          isDirty: true,
        },
      };
    case 'VALIDATION':
      return {
        ...state,
        errors: {
          ...state.errors,
          [field]: error,
        },
        meta: {
          ...state.meta,
          invalid: true,
        },
      };
    case 'FORM_VALIDATION':
      return {
        ...state,
        errors: payload,
        meta: {
          ...state.meta,
          invalid: true,
        },
      };
    case 'START_SUBMIT':
      return {
        ...state,
        meta: {
          ...state.meta,
          submitting: true,
        },
      };
    case 'STOP_SUBMIT':
      return {
        ...state,
        meta: {
          ...state.meta,
          submitting: false,
        },
      };
    case 'SUBMIT_SUCCEEDED':
      return {
        ...state,
        meta: {
          ...state.meta,
          submitting: false,
          submitSucceeded: true,
        },
      };
    case 'SUBMIT_FAILED':
      return {
        ...state,
        errors: payload,
        meta: {
          ...state.meta,
          submitting: false,
          submitSucceeded: false,
        },
      };
    case 'REGISTER_FIELD':
      if (isUndefined(get(state.values, payload)))
        set(state.values, payload, null);
      return state;
    case 'UNREGISTER_FIELD':
      unset(state.values, payload);
      return state;
    case 'INITIALIZE':
      return {
        ...initialFormReducerState,
        values: payload,
      };
    default:
      return state;
  }
};

function Form({
  children,
  initial = {},
  onSubmit,
  validate: validation,
  effect,
}) {
  const [state, dispatch] = useReducer(reducer, initialFormReducerState, {
    type: 'INITIALIZE',
    payload: initial,
  });

  const registerField = name =>
    dispatch({
      type: 'REGISTER_FIELD',
      payload: name,
    });

  const unregisterField = name =>
    dispatch({
      type: 'UNREGISTER_FIELD',
      payload: name,
    });

  const change = ({ field, value }) =>
    dispatch({
      type: 'CHANGE',
      payload: { field, value },
    });

  const validate = ({ field, error }) =>
    dispatch({ type: 'VALIDATION', payload: { field, error } });

  const submit = async () => {
    dispatch({
      type: 'START_SUBMIT',
    });

    const errors = validation(state.values);

    if (errors && !isEmpty(errors)) {
      dispatch({ type: 'FORM_VALIDATION', payload: errors });
      dispatch({
        type: 'SUBMIT_FAILED',
        payload: errors,
      });
    } else {
      try {
        await onSubmit(state.values);
        dispatch({
          type: 'SUBMIT_SUCCEEDED',
        });
      } catch (e) {
        dispatch({
          type: 'SUBMIT_FAILED',
          payload: e,
        });
      } finally {
        dispatch({
          type: 'SUBMIT_STOP',
        });
      }
    }
  };

  useEffect(() => dispatch({ type: 'INITIALIZE', payload: initial }), [
    initial,
  ]);

  useEffect(
    () => {
      effect({ state, dispatch, change });
    },
    [state],
  );

  return (
    <FormProvider
      value={{
        change,
        validate,
        state,
        submit,
        values: state.values,
        registerField,
        unregisterField,
      }}
    >
      {children}
    </FormProvider>
  );
}

Form.defaultProps = {
  initial: {},
  onSubmit: () => {},
  validate: () => {},
  effect: () => {},
  children: [],
};

Form.propTypes = {
  initial: shape({}),
  onSubmit: func,
  validate: func,
  effect: func,
  children: arrayOf(oneOfType([node, func, string])),
};

export default Form;
