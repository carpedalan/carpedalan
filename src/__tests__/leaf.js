import React from 'react';
import { create } from 'react-test-renderer';

import Form from '../form/Form';
import Leaf from '../leaf';

describe('test', () => {
  it('should do some stuff', () => {
    const instance = create(<Leaf />);
    expect(instance.toJSON()).toMatchSnapshot();
  });

  it('tests state changes', () => {
    const { root } = create(<Leaf />);

    root.findByProps({ 'data-test': 'clicky' }).props.onClick();
    expect(root.findAllByType(Form)[0].props.initial.test.yo).toBe('fartttsss');
  });
});
