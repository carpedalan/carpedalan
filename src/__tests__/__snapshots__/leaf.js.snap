// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should do some stuff 1`] = `
Array [
  <input
    onChange={[Function]}
    type="file"
  />,
  <input
    data-test="inputField"
    onChange={[Function]}
    value=""
  />,
  <div
    onClick={[Function]}
  >
    submit div
  </div>,
]
`;
